pkg/lib/import-json.ts

```diff
@@ -0,0 +1,182 @@
+ */
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string): never {
+    throw new ValidationError(`JSON validation error for ${validation_path.join("")}: ${msg}`);
+}
+
+export function import_string(val: JsonValue): string {
+    if (typeof val == "string")
+        return val;
+    validation_error(`Not a string: ${JSON.stringify(val)}`);
+}
+
+export function import_number(val: JsonValue): number {
+    if (typeof val == "number")
+        return val;
+    validation_error(`Not a number: ${JSON.stringify(val)}`);
+}
+
+export function import_boolean(val: JsonValue): boolean {
+    if (typeof val == "boolean")
+        return val;
+    validation_error(`Not a boolean: ${JSON.stringify(val)}`);
+}
+
+function is_json_object(val: JsonValue): val is JsonObject {
+    return !!val && typeof val == "object" && !Array.isArray(val);
+}
+
+export function import_json_object(val: JsonValue): JsonObject {
+    if (is_json_object(val))
+        return val;
+    validation_error(`Not an object: ${JSON.stringify(val)}`);
+}
+
+function is_json_array(val: JsonValue): val is JsonValue[] {
+    return !!val && typeof val == "object" && Array.isArray(val);
+}
+
+export function import_json_array(val: JsonValue): JsonValue[] {
+    if (is_json_array(val))
+        return val;
+    validation_error(`Not an array: ${JSON.stringify(val)}`);
+}
+
+export function import_record<T>(val: JsonValue, importer: (val: JsonValue) => T): Record<string, T> {
+    const obj = import_json_object(val);
+    return Object.fromEntries(Object.entries(obj).map(
+        ([k, v]) => [k, with_validation_path(`.${k}`, () => importer(v))]));
+}
+
+export function import_array<T>(val: JsonValue, importer: (val: JsonValue) => T): Array<T> {
+    const arr = import_json_array(val);
+    return arr.map((elt, i) => with_validation_path(`[${i}]`, () => importer(elt)));
+}
+
+export function import_optional<T>(obj: JsonObject, field: string, importer: (val: JsonValue) => T): T | undefined {
+    if (field in obj)
+        return with_validation_path(`.${String(field)}`, () => importer(obj[field]));
+    return undefined;
+}
+
+export function import_mandatory<T>(obj: JsonObject, field: string, importer: (val: JsonValue) => T): T {
+    if (field in obj)
+        return with_validation_path(`.${String(field)}`, () => importer(obj[field]));
+    validation_error(`Field "${String(field)}" is missing`);
+}
+
+export function validate<T>(path: string, val: JsonValue | undefined, importer: (val: JsonValue) => T, fallback: T): T {
+    if (val === undefined)
+        return fallback;
+
+    try {
+        return with_validation_path(path, () => importer(val));
+    } catch (e) {
+        if (!(e instanceof ValidationError))
```

Comment:
This block would definitely benefit from a comment explaining the two cases of errors that you intend to handle here and why they ought to be treated differently... I'm reasonably sure I understand it, but only "reasonably" :)

Reply:
It might also read better if you inverted the logic: ie: handle `if (e instanceof ValidationError)` in a block and then have `throw e` as a trailing/`else`.  That would make it easier to see that you're basically just doing `catch e: ValidationError` except that JS doesn't have that feature...

Reply:
Done.